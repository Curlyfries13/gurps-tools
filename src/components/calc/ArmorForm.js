import React, { useState } from 'react';
import { Collapse } from 'reactstrap';

import TextInput from '../common/TextInput';
import DRField from './DRField';

const ArmorForm = ({ armor, removeArmor }) => {
  const [isOpen, setIsOpen] = useState(true);
  const toggle = () => setIsOpen(!isOpen);

  return (
    <div className='container card card-body'>
      <div className='row g-2 mb-2'>
        <div className='col'>
          <button className='btn btn-primary' type='button' onClick={toggle}>
            {isOpen ? 'Collapse ' + armor.name : 'Open ' + armor.name}
          </button>
        </div>
        <div className='d-grid col-3 col-md2'>
          <button
            className='btn btn-outline-danger'
            type='button'
            onClick={() => removeArmor(armor)}
          >
            <i className='bi bi-shield-slash'></i>
          </button>
        </div>
      </div>
      <Collapse className='row' isOpen={isOpen}>
        <DRField armorId={armor.id} />
      </Collapse>
    </div>
  );
};

export default ArmorForm;
