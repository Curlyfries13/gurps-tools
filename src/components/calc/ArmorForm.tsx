import React, { useState, useEffect } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Subject } from 'rxjs';

import { Armor, MoveDirection } from 'src/types';
import { moveArmor, updateArmor } from 'src/redux/actions/armorActions';
import { RootState } from 'src/redux';

import { default as Collapse } from 'src/components/common/Collapse';
import TextInput from '../common/TextInput';
import DRField from './DRField';
import AblateField from './AblateField';

const ArmorForm = ({
  armor,
  removeArmor,
  collapse,
  moveArmor,
  updateArmor,
  armorStack,
}: Props) => {
  const [isOpen, setIsOpen] = useState(true);

  useEffect(() => {
    collapse.asObservable().subscribe({
      next: () => {
        setIsOpen(false);
      },
    });
    return function cleanup() {
      collapse.unsubscribe();
    };
  }, [collapse]);

  const toggle = () => setIsOpen(!isOpen);

  const handleMove = (direction: MoveDirection) => {
    moveArmor(armor, direction);
  };

  const handleAblateToggle = () => {
    updateArmor({ ...armor, ablative: !armor.ablative });
  };

  function handleAblateBaseUpdate(
    event: React.ChangeEvent<HTMLInputElement>
  ): void {
    const { name, value } = event.target;
    const base: number = parseInt(value, 10) || 0;
    const updatedArmor: Armor = { ...armor, ablateBase: base };
    updateArmor(updatedArmor);
  }

  function handleNameUpdate(event: React.ChangeEvent<HTMLInputElement>): void {
    const { name, value } = event.target;
    const updatedArmor: Armor = { ...armor };
    updatedArmor.name = value;
    updateArmor(updatedArmor);
  }

  const switchId = armor.id + '-ablative';

  return (
    <div
      className='container mb-2 card card-body'
      role='listitem'
      aria-label={armor.name + ' section'}
    >
      <div className='row g-2 mb-2'>
        <div className='col'>
          <button className='btn btn-primary' type='button' onClick={toggle}>
            {isOpen ? 'Collapse ' + armor.name : 'Open ' + armor.name}
          </button>
        </div>
        <div className='d-grid col-3 col-md2'>
          <button
            className='btn btn-outline-danger'
            type='button'
            onClick={() => removeArmor(armor)}
          >
            <i className='bi bi-shield-slash'></i>
          </button>
        </div>
      </div>
      <Collapse
        className='row md-flex justify-content-between b-2 mb-3'
        role='group'
        aria-label={'armor ' + armor.order + ' collapse group'}
        isOpen={isOpen}
      >
        <div className='col col-6'>
          <TextInput
            id={armor.id + '-name'}
            label='Name'
            value={armor.name}
            onChange={handleNameUpdate}
          />
          <DRField armorId={armor.id} />
        </div>
        <div className='col col-6'>
          <div className='row mb-1'>
            <div className='form-check form-switch'>
              <input
                className='form-check-input'
                onChange={() => handleAblateToggle()}
                checked={armor.ablative}
                type='checkbox'
                id={switchId}
              />
              <label className='form-check-label' htmlFor={switchId}>
                Ablative
              </label>
            </div>
          </div>
          <Collapse className='row mb-1' isOpen={armor.ablative}>
            <AblateField armorId={armor.id} />
          </Collapse>
        </div>
      </Collapse>
      <div className='row d-flex mb-1 g-4 justify-content-center'>
        <div className='d-grid col col-2'>
          <button
            className='btn btn-outline-secondary'
            onClick={() => handleMove(MoveDirection.DOWN)}
            disabled={armorStack.length === 1}
          >
            <i className='bi bi-caret-down-fill' />
          </button>
        </div>
        <div className='d-grid col col-2'>
          <button
            className='btn btn-outline-secondary'
            onClick={() => handleMove(MoveDirection.UP)}
            disabled={armorStack.length === 1}
          >
            <i className='bi bi-caret-up-fill' />
          </button>
        </div>
      </div>
    </div>
  );
};

interface OwnProps {
  armor: Armor;
  removeArmor: Function;
  collapse: Subject<void>;
}

function mapStateToProps(state: RootState, ownProps: OwnProps) {
  return {
    armorStack: state.armorStack,
    armor: ownProps.armor,
    removeArmor: ownProps.removeArmor,
  };
}

const mapDispatchToProps = {
  moveArmor,
  updateArmor,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type Props = ConnectedProps<typeof connector> & OwnProps;

export default connector(ArmorForm);
