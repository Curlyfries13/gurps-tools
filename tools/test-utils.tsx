// Using code from https://redux.js.org/recipes/writing-tests#connected-components
import React from 'react';

// test-utils.js
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
// Import your own reducer
//
// console.log(React.createElement);

import * as reducer from 'src/redux/reducers';


const customRender = (
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) => {
  const Wrapper =({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
