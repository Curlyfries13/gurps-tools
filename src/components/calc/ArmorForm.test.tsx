import React from 'react';
import { Subject } from 'rxjs';
import { RenderResult } from '@testing-library/react';
import { render } from 'tools/test-utils';

import ArmorForm from './ArmorForm';

const mockArmor = { dr: 0, id: 0 };
const mockState = { armorStack: [mockArmor] };
const mockCollapse = new Subject<void>();

it('matches snapshot', async () => {
  const component: RenderResult = render(
    <ArmorForm armor={mockArmor} collapse={mockCollapse} />,
    {
      initialState: mockState,
    }
  );
  expect(component).toMatchSnapshot();
});
