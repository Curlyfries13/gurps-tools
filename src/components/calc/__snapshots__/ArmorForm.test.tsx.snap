// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-label="undefined section"
        class="container mb-2 card card-body"
        role="listitem"
      >
        <div
          class="row g-2 mb-2"
        >
          <div
            class="col"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Collapse undefined
            </button>
          </div>
          <div
            class="d-grid col-3 col-md2"
          >
            <button
              class="btn btn-outline-danger"
              type="button"
            >
              <i
                class="bi bi-shield-slash"
              />
            </button>
          </div>
        </div>
        <div
          aria-hidden="false"
          aria-label="armor undefined collapse group"
          class="rah-static rah-static--height-auto"
          role="group"
          style="height: auto; overflow: visible;"
        >
          <div
            class="row md-flex justify-content-between b-2 mb-3"
          >
            <div
              class="col col-6"
            >
              <label
                class="form-label"
                for="0-name"
              >
                Name
              </label>
              <input
                aria-label="Name"
                class="form-control"
                id="0-name"
                type="text"
                value=""
              />
              <label
                class="col col-form-label d-flex justify-content-center"
              >
                DR
              </label>
              <label
                class="form-label d-flex justify-content-center"
                for=""
              />
              <input
                aria-label="Max Damage Resistence"
                class="form-control"
                id=""
                type="text"
                value="undefined"
              />
            </div>
            <div
              class="col col-6"
            >
              <div
                class="row mb-1"
              >
                <div
                  class="form-check form-switch"
                >
                  <input
                    class="form-check-input"
                    id="0-ablative"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="0-ablative"
                  >
                    Ablative
                  </label>
                </div>
              </div>
              <div
                aria-hidden="true"
                aria-label=""
                class="rah-static rah-static--height-zero"
                role=""
                style="height: 0px; overflow: hidden;"
              >
                <div
                  class="row mb-1"
                  style="display: none;"
                >
                  <div
                    class="row"
                  >
                    <label
                      class="col col-form-label"
                      for="0-ablateBase"
                    >
                      Ablate Base
                    </label>
                    <div
                      class="col"
                    >
                      <input
                        aria-label="this armor's ablative base"
                        class="form-control"
                        id="0-ablateBase"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row d-flex mb-1 g-4 justify-content-center"
        >
          <div
            class="d-grid col col-2"
          >
            <button
              class="btn btn-outline-secondary"
              disabled=""
            >
              <i
                class="bi bi-caret-down-fill"
              />
            </button>
          </div>
          <div
            class="d-grid col col-2"
          >
            <button
              class="btn btn-outline-secondary"
              disabled=""
            >
              <i
                class="bi bi-caret-up-fill"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="undefined section"
      class="container mb-2 card card-body"
      role="listitem"
    >
      <div
        class="row g-2 mb-2"
      >
        <div
          class="col"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Collapse undefined
          </button>
        </div>
        <div
          class="d-grid col-3 col-md2"
        >
          <button
            class="btn btn-outline-danger"
            type="button"
          >
            <i
              class="bi bi-shield-slash"
            />
          </button>
        </div>
      </div>
      <div
        aria-hidden="false"
        aria-label="armor undefined collapse group"
        class="rah-static rah-static--height-auto"
        role="group"
        style="height: auto; overflow: visible;"
      >
        <div
          class="row md-flex justify-content-between b-2 mb-3"
        >
          <div
            class="col col-6"
          >
            <label
              class="form-label"
              for="0-name"
            >
              Name
            </label>
            <input
              aria-label="Name"
              class="form-control"
              id="0-name"
              type="text"
              value=""
            />
            <label
              class="col col-form-label d-flex justify-content-center"
            >
              DR
            </label>
            <label
              class="form-label d-flex justify-content-center"
              for=""
            />
            <input
              aria-label="Max Damage Resistence"
              class="form-control"
              id=""
              type="text"
              value="undefined"
            />
          </div>
          <div
            class="col col-6"
          >
            <div
              class="row mb-1"
            >
              <div
                class="form-check form-switch"
              >
                <input
                  class="form-check-input"
                  id="0-ablative"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="0-ablative"
                >
                  Ablative
                </label>
              </div>
            </div>
            <div
              aria-hidden="true"
              aria-label=""
              class="rah-static rah-static--height-zero"
              role=""
              style="height: 0px; overflow: hidden;"
            >
              <div
                class="row mb-1"
                style="display: none;"
              >
                <div
                  class="row"
                >
                  <label
                    class="col col-form-label"
                    for="0-ablateBase"
                  >
                    Ablate Base
                  </label>
                  <div
                    class="col"
                  >
                    <input
                      aria-label="this armor's ablative base"
                      class="form-control"
                      id="0-ablateBase"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row d-flex mb-1 g-4 justify-content-center"
      >
        <div
          class="d-grid col col-2"
        >
          <button
            class="btn btn-outline-secondary"
            disabled=""
          >
            <i
              class="bi bi-caret-down-fill"
            />
          </button>
        </div>
        <div
          class="d-grid col col-2"
        >
          <button
            class="btn btn-outline-secondary"
            disabled=""
          >
            <i
              class="bi bi-caret-up-fill"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
